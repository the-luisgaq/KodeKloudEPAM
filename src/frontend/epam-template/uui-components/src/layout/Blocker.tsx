import React, { forwardRef } from 'react';
import { CSSTransition } from 'react-transition-group';
import { IHasCX, cx, IHasRawProps } from '@epam/uui-core';
import css from './Blocker.module.scss';

export interface BlockerProps extends IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    /** Turns the blocker on or off */
    isEnabled: boolean;
    /** Disables spinner animation display */
    hideSpinner?: boolean;
    /** Sets the minimal height in px, for cases when blocked content is empty */
    spacerHeight?: number;
    /** Replaces default spinner */
    renderSpinner?(props: any): React.ReactNode;
}

const classNames = {
    container: 'uui-blocker-container',
    blocker: 'uui-blocker',
};

const uuiBlocker = {
    enter: 'uui-blocker-enter',
    enterActive: 'uui-blocker-enter-active',
    exit: 'uui-blocker-exit',
    exitActive: 'uui-blocker-exit-active',
} as const;

export const Blocker = forwardRef<HTMLDivElement, BlockerProps>((props, ref) => {
    const transitionRef = React.createRef<HTMLDivElement>();
    return (
        <div
            className={ cx(css.container, classNames.container, props.cx) }
            style={ { minHeight: props.isEnabled && props.spacerHeight ? `${props.spacerHeight}px` : undefined } }
            ref={ ref }
            role="status"
            aria-live="polite"
            aria-busy={ props.isEnabled ? 'true' : 'false' }
            aria-label={ props.isEnabled ? 'Loading' : '' }
            { ...props.rawProps }
        >
            <CSSTransition
                in={ props.isEnabled }
                nodeRef={ transitionRef }
                classNames={ uuiBlocker }
                timeout={ { enter: 2000, exit: 1000 } }
                mountOnEnter
                unmountOnExit
            >
                <div ref={ transitionRef } className={ classNames.blocker }>
                    {!props.hideSpinner && props.renderSpinner && props.renderSpinner(props)}
                </div>
            </CSSTransition>
        </div>
    );
});
