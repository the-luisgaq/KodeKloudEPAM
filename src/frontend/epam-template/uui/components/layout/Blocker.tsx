import { withMods } from '@epam/uui-core';
import { Blocker as uuiBlocker, BlockerProps } from '@epam/uui-components';
import css from './Blocker.module.scss';
import { settings } from '../../settings';

export const Blocker = withMods<BlockerProps, BlockerProps>(
    uuiBlocker,
    () => [css.root],
    () => ({ renderSpinner: settings.blocker.renderSpinner }),
);
