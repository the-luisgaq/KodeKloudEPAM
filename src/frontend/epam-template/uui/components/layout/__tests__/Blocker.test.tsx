import React from 'react';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';
import { Blocker } from '../Blocker';

describe('Blocker', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(<Blocker isEnabled />);
        expect(tree).toMatchSnapshot();
    });
});
