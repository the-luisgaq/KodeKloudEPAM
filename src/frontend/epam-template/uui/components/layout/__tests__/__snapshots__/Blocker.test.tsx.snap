// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blocker should be rendered correctly 1`] = `
<DocumentFragment>
  <div
    aria-busy="true"
    aria-label="Loading"
    aria-live="polite"
    class="container uui-blocker-container root"
    role="status"
  >
    <div
      class="uui-blocker"
    >
      <div
        class="container uui-spinner-container root uui-spinner"
      >
        <div
          class="uui-spinner-animation"
        >
          <div
            class="uui-spinner-dot uui-spinner-dot-1"
          />
          <div
            class="uui-spinner-dot uui-spinner-dot-2"
          />
          <div
            class="uui-spinner-dot uui-spinner-dot-3"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
